### PR TITLE
9243 westfeld add deployment docs pip

### PR DESCRIPTION
**Pull request by westfeld**
see https://twistedmatrix.com/trac/ticket/9243

### CODE REVIEW

File: src/twisted/newsfragments/9243.doc
Comment (exarkun) [inline src/twisted/newsfragments/9243.doc]: instead of "documentation" perhaps say `The "Writing a twistd plugin" howto`

File: docs/core/howto/tap.rst
Comment (exarkun) [inline docs/core/howto/tap.rst:317]: The existing formatting in this file is an artifact of a conversion tool from a different format.  There's no real reason to try to mimic it.  One blank line separating sections is plenty.  Content should also be formatted following http://rhodesmill.org/brandon/2012/one-sentence-per-line/

File: docs/core/howto/tap.rst
Comment (exarkun) [inline docs/core/howto/tap.rst]: titlecase section headings I think

File: docs/core/howto/tap.rst
Comment (exarkun) [inline docs/core/howto/tap.rst]: I think `setuptools` is preferred by everyone now?  It provides nice things like `find_packages`, anyway, and no one really wants to maintain a manual list of packages.

File: docs/core/howto/tap.rst
Comment (exarkun) [inline docs/core/howto/tap.rst:364]: Perhaps it would be worth highlighting the part of this that's actually relevant to twistd plugins... The `twisted.plugins` entry in `packages`?

Also, does this really work?  For a long time I've had the impression that it does not.  Instead, obscure alternate options like `py_modules` are required.

File: docs/core/howto/tap.rst
Comment (exarkun) [inline docs/core/howto/tap.rst]: Maybe the same license as Twisted here? :)

File: docs/core/howto/tap.rst
Comment (exarkun) [inline docs/core/howto/tap.rst]: wheels are the best, right?  Any reason to recommend sdist instead of wheel here?

File: docs/core/howto/tap.rst
Comment (exarkun) [inline docs/core/howto/tap.rst:320]: The directory layout for plugins is described in the plugins document.  If those docs aren't clear, perhaps they should be improved.  Then this document could refer to that one.

File: docs/core/howto/tap.rst
Comment (exarkun) [inline docs/core/howto/tap.rst]: This is a "dropin file", technically.  The things inside it are plugins.

**Review by exarkun — DISMISSED** [review]
Thanks.  I've left some comments inline.

File: docs/core/howto/tap.rst
Comment (glyph) [reply docs/core/howto/tap.rst]: My understanding is that the best practice is to consider the `sdist` the "root" thing, which you should always upload, and wheels are derivative artifacts of that.  The differences are mostly only relevant when it comes time to start building native extensions, of course.

Comment (glyph) [reply docs/core/howto/tap.rst]: (That said I *totally* just upload wheels to PyPI for a bunch of my smaller projects and direct people to git repos if they want to build from source.)

File: docs/core/howto/tap.rst
Comment (glyph) [reply docs/core/howto/tap.rst:364]: I am still waiting for someone with real distutils savvy to answer my question here: https://stackoverflow.com/questions/7275295/how-do-i-write-a-setup-py-for-a-twistd-twisted-plugin-that-works-with-setuptools

None of the answers seem really satisfying.

However, I just do the `packages` thing in the various packages I distribute nowadays - see for example https://github.com/glyph/txsni/blob/master/setup.py - and nobody seems to complain.  I don't know if this would cause issues if someone tried to build a Debian packages from these things or not, but increasingly it seems that building for distribution in the global package-manager-managed Linux site-packages is a Worst Practice.

File: docs/core/howto/tap.rst
Comment (westfeld) [reply docs/core/howto/tap.rst]: That is possible, however the unchanged section headings do not use titlecase. For consistency I did not use it, too.

File: docs/core/howto/tap.rst
Comment (westfeld) [reply docs/core/howto/tap.rst:317]: I see. I tried to stay in this style. I think changing this would be worth another ticket, as the diff will get quite hard to read with so many changes.

Comment (twm) [reply docs/core/howto/tap.rst:317]: I filed https://twistedmatrix.com/trac/ticket/9776 for this.

File: docs/core/howto/tap.rst
Comment (twm) [inline docs/core/howto/tap.rst]: `append()` doesn't return the list.

Try:

```python
packages=find_packages() + ['twisted.plugins'],
```

File: docs/core/howto/tap.rst
Comment (twm) [inline docs/core/howto/tap.rst]: Typo: "neccesary" &rarr; "necessary"

File: docs/core/howto/tap.rst
Comment (twm) [inline docs/core/howto/tap.rst]: `license` and `long_description` are technically optional, so I suggest dropping them from this example to reduce noise.

File: docs/core/howto/tap.rst
Comment (twm) [inline docs/core/howto/tap.rst:348]: There's no indication that this is `setup.py` in the text. Perhaps use the `:caption:` option to note that?

```rest
.. code-block:: python
    :caption: ``setup.py``
```

File: docs/core/howto/tap.rst
Comment (twm) [inline docs/core/howto/tap.rst]: This link leads to a semi-broken page (no CSS?). We should probably link to [PyPA's guide](https://packaging.python.org/tutorials/packaging-projects/), anyway.

File: docs/core/howto/tap.rst
Comment (twm) [inline docs/core/howto/tap.rst]: Since it is 2020 now, I suggest changing `python` to `python3`.

**Review by twm — CHANGES_REQUESTED** [review]
Thanks for this! There are still a few things that must be addressed before we can merge, however.

I noted a serious issue with the example `setup.py` &mdash; I don't think it actually works as written.

The rest of the comments are minor. In addition to those I noted inline, "Python" should really be capitalized everywhere it appears in prose (including the news fragment).

Thanks for contributing to Twisted!

File: docs/core/howto/tap.rst
Comment (westfeld) [reply docs/core/howto/tap.rst]: For me the link worked, so I left it in there but added the official Python packaging guide link, too.

**Review by twm — APPROVED** [review]
Thanks for sticking with this! I've added the PR to the merge queue.

**Comment by westfeld** [comment]
Thanks for finally accepting the PR. It took 3 years, but finally it worked out.

I still think that improving the docs is an important issue and I learned that outlining changes in the mailing list before writing a PR is the better way to go.

**Pull request merged**
